### DISCUSSION TITLE
Use case for midi controller with drum pads

### DISCUSSION THREAD

**Discussion by RK11111111111** [probonopd/MiniDexed#596] [Ideas]
I have this idea for midi controllers with drum pads.  For reference I have a PCR-m30 a 25 keyboard, with 8 drum pads, 4 knobs and 4 buttons.

What I want to do is map each pad to a different channel then load a different dex voice+bank for each channel/pad, as well as a voice for the keys.   I want to change the whole thing whenever I switch performances. 

The general idea is I want to play risers and effects as transitions between parts.

It seems like it should be possible as is, but I am having a bugger of a time getting it to happen.  I can get a pretty good performance.ini, but switching all the sudden my keys are playing what my drum pads were mapped to.

Any tips on how to get this going?  I've read the bits on PCC and I am still baffled.

I get the drum pads are beyond the scope of dexed, but extra points if it can make the after touch, resonance and filter cutoff and reverb to be Omni and work on all channels.

**Comment by diyelectromusic** [comment]
I'm not entirely clear what you're describing tbh... are you using 1 pad per MIDI channel - so 1 pad to control each tone generator?  Presumably you're reserving one channel for the voice, so am only using 7 of the pads?

Part of the problem is that the MIDI channels are included in the performance configurations so if you switch to a different performance and that has a different MIDI mapping it will change the MIDI channel settings to those in the new performance file.

To be able to select alternative performances, you'll need to send Program Change messages on the PCCH MIDI channel, but each individual tone generator will start listening to whatever MIDI channels are configured in the performance file you switch to.  This could result in stuck notes and similar if the MIDI channel changes between NoteOn and NoteOff messages (for example).

If you want to keep your per-pad MIDI configuration I'm afraid you'll have to ensure that each performance has the same MIDI configuration (your configuration) embedded in it.  This means setting up each of the: MIDIChannelX entries (where X goes from 1 to 8, one for each tone generator) to be the same in any performances you might end up selecting.

As an aside, and in an attempt to be able to suggest something a little easier for you to setup, it would appear that the properties in the performance files are recognised in the order they appear in the file, so if there is a duplicate the first instance seems to win out.

If so, then you could have a block at the top of each performance as follows:
MIDIChannel1=
MIDIChannel2=
MIDIChannel3=
MIDIChannel4=
MIDIChannel5=
MIDIChannel6=
MIDIChannel7=
MIDIChannel8=

With your multi tone-generator MIDI configuration in it and cut and paste that into the start of each .ini file in the performance directory and it will override any settings that appear further down in the files.

I can't guarantee this will always work, but at present this seems to be how the basic properties system works.

(I'm now minded to allow for a general "system configuration" properties file for things like MIDI channels that can be recognised for every performance - a bit like a #include in C/C++ or import in python...)

Another quirk - if the performance properties file includes a voice data block: VoiceData1= etc then this will always be the voice data loaded for that tone generator - in this case the voice/bank numbers are ignored... I don't know if that is relevant too?

Kevin

**Reply by RK11111111111** [reply]
So yes, numerically, the first pad also is sending notes on channel 1.  I put it in the bass range so it sounds different than the 25 keys.

So, I tried your suggestion.  I think I am getting somewhere.  I set aside all performances, except the default.  I then edited the ini to have the midi channel number be the same as the TG number.  I then when I booted I set each TG to have a different voice.  Oddly enough part of the way through, I lost the main keyboard altogether, though all of the pads worked as expected.  I went ahead and saved the ini, to see what was happening.  When I restarted the pi, the key bed function came back however the keys weren't playing the voice that was loaded in the first channel.  If I tried to edit the voice it went back to hopping around the TG when ever I hit a select or prev. button. CC113 and CC112 respectively.

**Reply by diyelectromusic** [reply]
Ah, in that case I suspect you're suffering from the fact that saving a performance embeds the currently active voice data in the performance.ini and from that point on will therefore ignore bank/voice numbers and only load in the raw voice data...  more on what's going on here: https://github.com/probonopd/MiniDexed/discussions/454#discussioncomment-5520090

I think this is a bug, but we are divided on the topic :)

Currently this is the only way to save a voice if you've been editing voice parameters...

By the way, on the "hopping" question - do you have a ButtonPinShortcut configured in minidexed.ini?  I've never used it but I think it is a quick way to edit the same parameter on several tone generators and might describe what you're seeing... but I can't say for sure how it is meant to work - as I say, I don't use it myself.

btw: note to self, the MIDI thing was discussed more here: https://github.com/probonopd/MiniDexed/pull/500#issuecomment-1675966197

Kevin

**Reply by RK11111111111** [reply]
I tried deleting the voice data of course, as I am suspect of these.  However, that doesn't make the situation better.  Just in case, I tried copying the voice data over from another performance, no luck.

The thing is that it isn't even playing that voice data with the key bed, the pad on channel one was playing the correct voice.  It's like it split the channel, is minidexed trying to send commands to the keyboard to change channel?

**Comment by Banana71** [comment]
From RK111111111:
> Yeah it's a mess. With my configuration, I have 8 midi channels mapped. One through eight in my keyboard, 7map to pads and one to the keybed, and a pad(there are 8 pads). In the performances.ini default, I set each TG to a different channel one through eight. I have for midi momentary buttons two for next and back. One for previous and one for select. When I hit select it toggles the keybed of the device to different midi channels I can step through each one just by hitting select and then back through by hitting prev. It will go past 8 channels, but there is silence, but I can go back. However, using the forward and next buttons I can navigate the menu next goes from TG1 to TG2(but the sound stays the same.) I eventually hit the effects menu, then the performance menu. If I hit select and go down to load, I can load another performance. Interestingly enough, it loads all the TG correctly, but the keybed is playing the third TG(TG3), then I can hit prev and go to the top level menu, then I can play the performance as specified in the performance.

> I say this, because while it is a bug, it is kind of a cool bug to be able to go through the TG as specified in the performances. Though I have a hard time explaining it to my daughter. Ideally, IMO, it would be the next and back keys, when the TG2 is displayed, the second tone generator would make sound. It's pretty weird because it seems to be literally remapping the midi device driver to different midi channels, as my keyboard actually registers the pads as a separate device, and I set one pad to channel one which doesn't change. Weirdly enough, if I unplug/replug the keyboard USB the midi resets, and everything works as designed. I thought it was a rogue sysex midi send messing with the keyboard, but mini dexed uses a different sysex device ID. I also tested a number of different hardwiring of channels/tone generators, and it appears that something in dexed itself changes.

> I haven't got out my edirol pcr-m30 to test it, that device can actually store/edit different midi channel configurations and select different channels. I hesitate to use that keyboard, because mini-dexed has no use for all the sliders and knobs.

Hi @RK11111111111 
I don't understand exactly what you want and what you're doing. Or are your expectations not the right ones?

The following information would help me to help you: The Performance.ini, minidexed.ini and a picture from a record of which MIDI data is sent from the different PAD's, the keyboard and the navigation keys from your MIDI controller keyboard (e.g. PC program MIDI-OX ).

Example:
![Screenshot 2024-01-20 071836](https://github.com/probonopd/MiniDexed/assets/104296205/ae637e82-0026-4b29-a860-62a7e0f4b4a4)

Timestamp 08EA: PAD 1, should play TG1, but nothing can be hear
Timestamp 20BE: PAD 2, should play TG2, but TG3 can be hear
...


Peter
